runner.py

```diff
@@ -0,0 +1,285 @@
+
+        # If the return code is not 0, then it failed.
+        if code != 0:
+            self.attempts -= 1
+
+            if self.__sys_trace:
+                create_log_file('sys-trace', self.__captured_sys_trace)
+
+            if self.__call_trace:
+                create_log_file('call-trace', f'\n{stderr.decode()}')
+
+            # TODO: click.echo() should not be in the command class.
+            if self.__log_trace:
+                if stdout:
+                    click.echo(f'\n{stdout.decode()}')
+
+                if stderr:
+                    click.echo(f'\n{stderr.decode()}')
+
+        # Add or update the return code for the execution summary.
+        self.__return_codes[code] = self.__return_codes.get(code, 0) + 1
+
+    @property
+    def attempts(self):
+        """Returns the attempts."""
+        logging.debug('attempts.getter')
+        return self.__attempts
+
+    @attempts.setter
+    def attempts(self, value):
+        """Set the attempts value.
+
+        Parameters
+        ----------
+        value : int
+            The new value.
+        """
+        logging.debug('attempts.setter')
+        self.__attempts = value
+
+    def summary(self):
+        """Returns a summary of the command execution."""
+        logging.debug('summary()')
+
+        result = '--- command execution statistics ---'
+        codes = self.__return_codes
+
+        # If no return codes are available, return the result.
+        if len(codes) == 0:
+            logging.debug('summary() -> len(__return_codes) = 0')
+            return result
+
+        # Add all return codes to the result.
+        for key, value in codes.items():
+            result += (f'\nreturn code: {key} amount: {value}')
+
+        # Add the most frequent return code to the result.
+        result += f'\nmost frequent return code: {max(codes, key=codes.get)}'
+        return result
+
+
+def capture_sys_trace(process_id):
+    """Captures and returns the system trace for the specified process.
+
+    Parameters
+    ----------
+    process_id : int
+        The process identifier.
+    """
+    logging.debug('capture_sys_trace()')
+
+    # Captures the specified process.
+    process = psutil.Process(process_id)
+
+    # Receives a trace and returns it.
+    return [f'Disk I/O: {str(psutil.disk_io_counters(perdisk=False))}',
+            f'\n% Memory: {str(process.memory_percent())}',
+            f'\n% CPU: {str(process.cpu_percent())}',
+            f'\nThreads: {str(process.threads())}',
+            f'\nNetwork Counters: {str(psutil.net_io_counters())}']
+
+
+def create_log_file(name, content):
+    """Create a log file.
+
+    Parameters
+    ----------
+    name : str
+        Name of the log file.
+    content : str
+        Log file content.
+    """
+    logging.debug('create_log_file()')
+
+    # Set name for the log file.
+    name = f'{name}-{time.strftime("%H-%M-%S", time.localtime())}.log'
```

Comment:
the main point of this function is to generate a unique file name(which should be a seperate function regardless).
with context manager this function becomes much shorter and safer.(there are a lot of exceptions that you ignore)